openedx/core/djangoapps/catalog/docs/decisions/002-course-models.rst

```diff
@@ -0,0 +1,76 @@
+Course Models
+#############
+
+Status
+******
+
+Accepted
+
+Decision
+********
+
+edx-platform will have a first-class model to represent the idea of a catalog
+course. If we think of multiple runs of the same course to all derive from one
+conceptiual course.  The catalog course would be that conceptual course.
+
+Context
+*******
+
+Historically this concept has been inferred from the names of courses with the
+runs removed (eg. "AximX/Eng101" could be a catalog course with multiple runs
+(2024H1, 2024H2, etc)). Having this explicitly defined as a new model has
+multiple benefits.
+
+* Relationships between course runs can be defined explicitly instead of implicitly.
+
+  * Currently this information is either implied or we require the entire
+    course-discovery service in order to retrieve this information.
+
+* A list of all courses can be easily queried both internally and via the rest API.
+
+* Courses can be explicitly mapped to organizations rather than implicitly.
+
+* This change opens up the door for program or course bundle data to be persisted
+so that a reset of mem cache doesn't break programs related features.
+
+* Org → Course → Course Runs mappings don't have to follow implicit naming
+  Conventions and associations can be changed more easily.
+
+Rejected Alternatives
+*********************
+
+Rely on course-discovery for CatalogCourse
+==========================================
+
+The course-discovery service already has some models to represent this and we
+could rely on it directly but it also contains a lot of representations that
+are specific to the edx.org process and it's not valuable to spin up the
+course-discovery service just to get this information for most operators.
+
+As long as any local data can be updated from course-discovery if
+course-discovery is setup in an environment, we should not have any conflicts
+between the two locations.
+
+Essentially, if an operator has course-discovery enabled, the models here would
+essentially be a persisted cache of the relevant information.  However, if
+course-discovery is not enabled, the models can stand-alone and be populated
+either directly via APIs or automatically based on course naming conventions so
+that any features that rely on them can work in either system.
+
+
+Implementation Details
+**********************
+
+* Create a new CatalogCourse model
+
+* A 1:N mapping between 1 Catalog Course and N Course Overview objects.
```

Comment:
added a sub-bullet, as I think @ormsbee had some reasonable reservations about CourseOverview last time we chatted

```suggestion
* A 1:N mapping between 1 Catalog Course and N Course Run objects.

  * Open question: do we re-use CourseOverview to represent Course Run, or do we make a new CourseRun run model?
```

Reply:
To expand a bit on what @kdmccormick said:

The CourseOverview model is a giant hodgepodge of metadata attributes that has been added to over the years. There are also some weird attributes that still proxy back to the ModuleStore, and it has its own versioning/migration system that's always been a bit peculiar.

But my biggest concern about CourseOverviews and this new set of models is the data flow relationship. CourseOverviews are derived from Studio ModuleStore content. Some of the catalog use cases involve provisioning course runs and the roles associated with those runs before any content is even authored–which would suggest a data relationship where the CatalogCourse exists first, and we later make a relationship to a particular ModuleStore course run. But that raises a bunch of questions about what the source of truth actually is, and how data imported/exported from Studio interacts with catalog data that can also be altered outside the system.

So does Catalog know about Content, and say, auto-provision an empty course run when it needs to? But then there are things that are interested in the abstract notion of a catalog course and the roles associated with it, that have nothing to do with the content itself (e.g. enrollment reports). Does the relationship go the other way, where the content portion pushes data into the catalog course when things change? Do we remove some of that metadata from the OLX so that it becomes unversioned and separate from what's stored in the ModuleStore altogether? If we did, what would be the implications for import/export across instances. Etc.

I feel like that's the trickiest part about this proposal. The low-level data flow and code dependency issues that nobody's going to think about if we get it right, but will be a perpetual operational pain if we get it wrong.